torch/distributed/_tensor/_redistribute.py

```diff
@@ -97,43 +86,131 @@ def _gen_transform_infos(
+        # In those cases, we first tranverse from inner placement to outer placement
+        # to detect misaligned shardings and properly replicate nested sharding
+        search_transform_infos(
+            src_placements,
+            dst_placements,
+            device_mesh.ndim - 1,
+            device_mesh.ndim,
+            mesh_dims_to_logical_shape,
+            transform_infos,
+            left_to_right=False,
+        )
+
+    # we always tranverse from outer placement to inner placement to generate the transform infos
+    search_transform_infos(
+        src_placements,
+        dst_placements,
+        0,
+        device_mesh.ndim,
+        mesh_dims_to_logical_shape,
+        transform_infos,
+    )
+
+    return transform_infos
+
+
+def reshardable_from_src_to_dst(src_placements, dst_placements, mesh_dim) -> bool:
+    src = src_placements[mesh_dim]
+    dst = dst_placements[mesh_dim]
+
+    # For src that is Shard, check if it's the inner most sharding
+    # if not then it's not reshardable, else we check the dst sharding
+    if src.is_shard():
+        for i in reversed(range(len(src_placements))):
+            if src_placements[i].is_shard(src.dim):
+                if i != mesh_dim:
+                    return False
+                else:
+                    break
+
+    if not dst.is_shard():
+        return True
+
+    shard_dim = dst.dim
+    # For dst that is Shard, check if src/dst_placements have the same
+    # sharding on the tensor dim BEFORE the current mesh_dim
+    src_dim_count, dst_dim_count = 0, 0
+    for i, (s, p) in enumerate(zip(src_placements, dst_placements)):
+        if i >= mesh_dim:
+            break
+        # if isinstance(s, Shard) and s.dim == shard_dim:
+        if s.is_shard(shard_dim):
+            src_dim_count += 1
+        if p.is_shard(shard_dim):
+            dst_dim_count += 1
+
+    return src_dim_count == dst_dim_count
+
+
+def search_transform_infos(
+    src_placements: List[Placement],
+    dst_placements: List[Placement],
+    current_mesh_idx,
+    mesh_ndim,
+    mesh_dims_to_logical_shape,
+    transform_infos: List[_TransformInfo],
+    left_to_right=True,
+) -> None:
+    # short circuit
+    if src_placements == dst_placements:
+        return
+
+    if current_mesh_idx < 0 or current_mesh_idx >= mesh_ndim:
+        return
+
+    src = src_placements[current_mesh_idx]
+    dst = dst_placements[current_mesh_idx]
+
+    if reshardable_from_src_to_dst(src_placements, dst_placements, current_mesh_idx):
+        # can transform directly
+        if src != dst:
             transform_infos.append(
                 _TransformInfo(
-                    mesh_dim=i,
+                    mesh_dim=current_mesh_idx,
                     src_dst_placements=(src, dst),
-                    logical_shape=mesh_dims_to_logical_shape[i],
+                    logical_shape=mesh_dims_to_logical_shape[current_mesh_idx],
```

Comment:
When doing right to left pass, `mesh_dims_to_logical_shape` is not modified when `src` is shard and `dst` is replicate. Would this become an issue if we go from left to right, since `mesh_dims_to_logical_shape` would become obsolete?

Reply:
I don't think it would have any effect here, the reason is that `mesh_dims_to_logical_shape` is mainly used for replication (shard -> replicate) purpose where we need to know the logical shape when it's uneven sharded, so that it would be used when going from right to left, but not left to right.